### PR TITLE
Android permission check

### PR DESCRIPTION
**Pull request by ErnestG**
Added a simple permission check and request to prevent crashing on api 23+

### CODE REVIEW

**Comment by vrv** [comment]
ping for @andrewharp

**Comment by andrewharp** [comment]
Hi, thanks for the PR. We had previously merged a commit to add a permissions check (https://github.com/tensorflow/tensorflow/pull/926), but reverted it because at the time it was causing some Google-internal build complications.

Instead it was easier to change the app to build at API level 21, where all required permissions are granted on install automatically (are you trying to build at API 23? If so, maybe adb install -g will work for you in the short term?)

My major concern here is bringing in the Android support libraries as a dependency, since that complicates the build. I can test the previous PR again and see if that builds internally now, if this works for you.

**Comment by andrewharp** [comment]
I checked, and I can no longer reproduce the internal issue with the previous change.

Also, it does make more sense to put the authorization code in CameraActivity.java and to additionally request the storage permission, as you've done. Would you mind modifying your commit remove the use of the Android support libraries (you can use the previous solution for reference)?

**Comment by ErnestG** [comment]
@andrewharp made the change. Do you want to set the target at 23 in the same commit?

File: tensorflow/examples/android/src/org/tensorflow/demo/CameraActivity.java [hunk 1]
```diff
@@ -64,18 +63,24 @@ public void onRequestPermissionsResult(int requestCode, String permissions[], in
   }
 
   private boolean hasPermission() {
-    if (ContextCompat.checkSelfPermission(this, PERMISSION_CAMERA) == PackageManager.PERMISSION_DENIED || ContextCompat.checkSelfPermission(this, PERMISSION_STORAGE) == PackageManager.PERMISSION_DENIED) {
-      return false;
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+      if (checkSelfPermission(PERMISSION_CAMERA) == PackageManager.PERMISSION_DENIED || checkSelfPermission(PERMISSION_STORAGE) == PackageManager.PERMISSION_DENIED) {
```
Comment (andrewharp) [inline tensorflow/examples/android/src/org/tensorflow/demo/CameraActivity.java]: No need for the blank line, and you can also do:
`return checkSelfPermission(PERMISSION_CAMERA) != PackageManager.PERMISSION_DENIED && checkSelfPermission(PERMISSION_STORAGE) != PackageManager.PERMISSION_DENIED;`
to remove the branching here.

File: tensorflow/examples/android/src/org/tensorflow/demo/CameraActivity.java [hunk 2]
```diff
@@ -64,18 +63,24 @@ public void onRequestPermissionsResult(int requestCode, String permissions[], in
...
+      }
     } else {
       return true;
     }
   }
 
   private void getPermission() {
-    if (ActivityCompat.shouldShowRequestPermissionRationale(this, PERMISSION_CAMERA) || ActivityCompat.shouldShowRequestPermissionRationale(this, PERMISSION_STORAGE)) {
-      Toast.makeText(CameraActivity.this, "Camera AND storage permission are required for this demo", Toast.LENGTH_LONG).show();
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
```
Comment (andrewharp) [inline tensorflow/examples/android/src/org/tensorflow/demo/CameraActivity.java]: extra blank line here

File: tensorflow/examples/android/src/org/tensorflow/demo/CameraActivity.java [hunk 3]
```diff
@@ -64,18 +63,24 @@ public void onRequestPermissionsResult(int requestCode, String permissions[], in
...
+      if (shouldShowRequestPermissionRationale(PERMISSION_CAMERA) || shouldShowRequestPermissionRationale(PERMISSION_STORAGE)) {
+        Toast.makeText(CameraActivity.this, "Camera AND storage permission are required for this demo", Toast.LENGTH_LONG).show();
+      }
+      requestPermissions(new String[]{PERMISSION_CAMERA, PERMISSION_STORAGE}, PERMISSIONS_REQUEST);
```
Comment (andrewharp) [inline tensorflow/examples/android/src/org/tensorflow/demo/CameraActivity.java]: missing space after String[]

**Comment by andrewharp** [comment]
Thanks!

With the runtime checks on Android M for devices at API < 23, yes it should be safe (and required) to set the target level to 23.

**Comment by ErnestG** [comment]
@andrewharp let me know if everything looks good.

**Comment by andrewharp** [comment]
@tensorflow-jenkins: test this please

**Comment by vrv** [comment]
@andrewharp ready to merge?

**Pull request merged**